packages/frontend/core/src/components/attachment-viewer/utils.ts

```diff
@@ -0,0 +1,87 @@
+import type { AttachmentBlockModel } from '@blocksuite/affine/blocks';
+import { filesize } from 'filesize';
+
+import { downloadBlob } from '../../utils/resource';
+
+export async function getAttachmentBlob(model: AttachmentBlockModel) {
+  const sourceId = model.sourceId;
+  if (!sourceId) {
+    return null;
+  }
+
+  const doc = model.doc;
+  let blob = await doc.blobSync.get(sourceId);
+
+  if (blob) {
+    blob = new Blob([blob], { type: model.type });
+  }
+
+  return blob;
+}
+
+export async function download(model: AttachmentBlockModel) {
+  const blob = await getAttachmentBlob(model);
+  if (!blob) return;
+
+  await downloadBlob(blob, model.name);
+}
+
+export function buildAttachmentProps(model: AttachmentBlockModel) {
+  const isPDF = model.type.endsWith('pdf');
+  const pieces = model.name.split('.');
+  const ext = pieces.pop() || '';
+  const name = pieces.join('.');
+  const size = filesize(model.size);
+  return { model, name, ext, size, isPDF };
+}
+
+/**
+ * Generates a set of sequences.
+ *
+ * 1. when `start` is `0`, returns `[0, .., 5]`
+ * 2. when `end` is `total - 1`, returns `[total - 1, .., total - 5]`
+ * 3. when `start > 0` and `end < total - 1`, returns `[18, 17, 19, 16, 20, 15, 21]`
+ */
+export function genSeq(start: number, end: number, total: number) {
```

Comment:
no longer used

Reply:
may need it when scrolling through long lists

Reply:
page rendering is controlled by virtuoso, all we need to do is cancel previous rendering properly. no more hack

Reply:
https://virtuoso.dev/scroll-seek-placeholders/

Reply:
it requires range not sequences. this function is only useful when handy controlling page visibilities